include/nut_float.h

```diff
@@ -0,0 +1,48 @@
+/*
+ * nut_float.h - Network UPS Tools include files for floating point routines
+ * and ways to compare if non-integers are close enough to assume equal
+ *
+ * Copyright (C) 2011	Arjen de Korte <adkorte-guest@alioth.debian.org>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#ifndef NUT_FLOAT_H_SEEN
+#define NUT_FLOAT_H_SEEN 1
+
+#include "config.h"
+
+#if defined HAVE_FLOAT_H
+#  include <float.h>
+#endif
+
+#if defined HAVE_MATH_H
+#  include <math.h>
+#endif
+
+/* Modern compilers frown upon direct comparisons of floating point numbers
+ * since their imprecise internal representations can misfire, and really we
+ * care if they are indiscernably close. To aviod warnings like -Wfloat-equal
+ * we prefer to compare with methods defined below. Note: despite the "floating"
+ * name, fabs() seems to be defined over doubles or wider types, e.g.:
+ *     double fabs(double x);
+ *     float fabsf(float x);
+ *     long double fabsl(long double x);
+ */
+#define f_equal(x, y) ( fabsf((float)(x) - (float)(y)) < FLT_EPSILON )
```

Comment:
Should this be `< 2*FLT_EPSILON` or `<= FLT_EPSILON`? Otherwise, if epsilon is the closest number to zero, the only way the delta can be less than epsilon is if it is exactly 0 (which is the case the compiler is complaining about).

Reply:
Interesting consideration. I don't think any discussions on float comparisons I saw in the research leading up to this solution went into this detail, although there were quite a few examples with strict "less". To err on side of caution, might go with "small difference still means zero for our needs" :)

Per various layman definitions in headers etc., these epsilons are defined like:
* "The macro yields the smallest X of type float such that 1.0 + X != 1.0." - https://www.qnx.com/developers/docs/6.4.1/dinkum_en/cpp/float.html
* "This is the difference between 1 and the smallest floating point number of type float that is greater than 1. It’s supposed to be no greater than 1E-5." (1E-9 for double and long double) - https://www.gnu.org/software/libc/manual/html_node/Floating-Point-Parameters.html
* "difference between 1.0 and the next representable value for float, double and long double respectively" - https://en.cppreference.com/w/c/types/limits

So it is probably not exactly the minimum representable value in the bits of the discrete floating point numeric type, I don't see a direct correlation for comparisons to zero. Got me wondering if the math should be like `fabs( x - y ) + 1.0 > 1.0 + FLT_EPSILON` or some such :)

Reply:
> Got me wondering if the math should be...

I think adding 1.0 to a very small number makes the comparison problem worse.

The `<= *_EPSILON` comparison should keep future compilers quiet (and in most of these cases, NUT was well-served by the original round-to-int semantics anyway).